### ISSUE TITLE
Parsing ambiguity with numeric literals and calling methods

### ISSUE THREAD

**Issue by baronfel** [dotnet/fsharp#12516]
When calling methods on integer numeric literals (and possible other forms of literals that may be ambiguous), the user is presented with an error about a malformed numeric literal.  While correct from the compiler's perspective, this appears trivially solvable to the user, so we should investigate changes to the error message or parser/checker enhancements to make this form understandable to the compiler.

**Repro steps**

<details>
<summary>Sample code where an extension method is added to the `int` type and then invoked on the `1` literal, causing a parsing error</summary>

```fsharp
open System.Runtime.CompilerServices

[<Extension>]
type StringExtensions = 
    [<Extension>]
    static member A(x: int) = 0
    [<Extension>]
    static member A(x: float) = 0
    [<Extension>]
    static member A(x: bool) = 0
    
let intExtension = 1.A() // error on this line
let floatExt = 1.0.A()
let boolExt = true.A()
```
</details>

**Expected behavior**

The compiler should be able to interpret the right hand side of `intExtension` as `(1).A()`, that is calling the member `A` on the value `1`.

**Actual behavior**

The compiler tries to parse the numeric literal `1.A`, can't, and generates the following error:

```
error FS1156: This is not a valid numeric literal. Valid numeric literals include 1, 0x1, 0o1, 0b1, 1l (int), 1u (uint32), 1L (int64), 1UL (uint64), 1s (int16), 1y (sbyte), 1uy (byte), 1.0 (float), 1.0f (float32), 1.0m (decimal), 1I (BigInteger).
```

**Known workarounds**

Add parenthesis around the `1` literal to allow the compiler to parse the numeric literal before going ahead with method invocation.

**Related information**

* [Sharplab link](https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAOgCUBXJbASzwQIGEY84LgEAnNNgNwoGMEIAsAChhAbQA8AUQAeuJBAowkAPgC6w7BkTpsrCkgDmMuQqUQUAXhTCUtlBOPJTK9ULsocAQ0o8U1PABGbCgAggAU0iAo+tgAlJYoAAw2dg6yToouKbZePn74QayhEVFgwDDe8VbJbqlS6fKZatke2N68+YHB4ZEoATAwwFVJLcIs2NHkjo1KCQCMBOGxYwgTZRXYxvMEiYthy0LjfQPAW1a6JDRLQA==)
* [Twitter thread](https://twitter.com/ChetHusk/status/1471147557447475208)

cc @KathleenDollard  as co-discoverer

**Comment by dsyme** [comment]
Yes from the language design perspective this is by design, since `1.` is a valid floating-point numeric literal, and so there's never any real way we're going to accept `1.ToString()` and so on.

Agreed diagnostics can be better here, also we shouldn't be offering auto-completes in such situations (I presume we aren't but there may be corner cases)

**Comment by baronfel** [comment]
It feels like there could be some limited form of lexer error recovery for this specific case in the lexer where the error is thrown:

https://github.com/dotnet/fsharp/blob/85fe2809db9c106d3c630c74dbe806eb8b9b09c9/src/fsharp/lex.fsl#L521-L522

Could we alter the emitted token or do something in the lex filter to suppress the initial error until type checking, then if no such method was found emit the numeric literal error?
